File: fog-layer/tsdbmanager.py
# tsdb_manager.py - Gestor de Base de Datos de Series Temporales
import os
from influxdb_client import InfluxDBClient, Point
from influxdb_client.client.write_api import SYNCHRONOUS
from dotenv import load_dotenv

# Cargar variables de entorno desde .env
load_dotenv()

class TimeSeriesManager:
    def __init__(self):
        self.token = os.getenv("INFLUXDB_TOKEN")
        self.org = os.getenv("INFLUXDB_ORG", "transwatch")
        self.bucket = os.getenv("INFLUXDB_BUCKET", "parking_data")
        self.url = os.getenv("INFLUXDB_URL", "http://localhost:8086")

        # Inicializar cliente de InfluxDB
        self.client = InfluxDBClient(url=self.url, token=self.token, org=self.org)
        self.write_api = self.client.write_api(write_options=SYNCHRONOUS)
        self.query_api = self.client.query_api()

    def almacenar_lectura(self, datos, device_id, qc_status):
        try:
            # Crear el punto de medición
            point = (
                Point("sensor_reading")
                .tag("device_id", device_id)
                .tag("qc_status", str(qc_status))
                .field("temp_celsius", float(datos.get("temp_celsius", 0.0)))
                .field("humedad_porcentaje", float(datos.get("humedad_porcentaje", 0.0)))
                .field("luz_adc", float(datos.get("luz_adc", 0.0)))
                .field("distancia_cm", float(datos.get("distancia_cm", 0.0)))
                .field("vehiculo_en_entrada_detectado", bool(datos.get("vehiculo_en_entrada_detectado", False)))
                .field("barrera_abierta", bool(datos.get("barrera_abierta", False)))
                .field("luces_parking_encendidas", bool(datos.get("luces_parking_encendidas", False)))
                .field("alarma_temperatura_activa", bool(datos.get("alarma_temperatura_activa", False)))
            )

            # Escribir en InfluxDB
            self.write_api.write(bucket=self.bucket, org=self.org, record=point)
            print(f"✅ Datos almacenados en InfluxDB para dispositivo {device_id}")
            return True

        except Exception as e:
            print(f"❌ Error almacenando en InfluxDB: {e}")
            return False


# Ejemplo de uso
if __name__ == "__main__":
    manager = TimeSeriesManager()

    datos_sensor = {
        "temp_celsius": 28.5,
        "humedad_porcentaje": 63.2,
        "luz_adc": 512,
        "distancia_cm": 85.4,
        "vehiculo_en_entrada_detectado": False,
        "barrera_abierta": True,
        "luces_parking_encendidas": True,
        "alarma_temperatura_activa": False
    }

    manager.almacenar_lectura(datos_sensor, device_id="sensor_A1", qc_status=True)

File: fog-layer/tsdbmanager.js
// tsdbManager.js - Gestor de Base de Datos de Series Temporales (versión JavaScript)
import 'dotenv/config';
import { InfluxDB, Point } from '@influxdata/influxdb-client';

export class TimeSeriesManager {
  constructor() {
    this.token = process.env.INFLUXDB_TOKEN;
    this.org = process.env.INFLUXDB_ORG || 'transwatch';
    this.bucket = process.env.INFLUXDB_BUCKET || 'parking_data';
    this.url = process.env.INFLUXDB_URL || 'http://localhost:8086';

    this.client = new InfluxDB({ url: this.url, token: this.token });
    this.writeApi = this.client.getWriteApi(this.org, this.bucket, 'ns');
    this.queryApi = this.client.getQueryApi(this.org);
  }

  async almacenarLectura(datos, device_id, qc_status) {
    // Crea un punto para InfluxDB
    const point = new Point('sensor_reading')
      .tag('device_id', device_id)
      .tag('qc_status', qc_status)
      .floatField('temperature', datos.temperatura)
      .floatField('humidity', datos.humedad)
      .floatField('light_level', datos.luz_adc)
      .floatField('distance', datos.distancia_cm)
      .booleanField('vehicle_detected', datos.vehiculo_en_entrada_detectado)
      .booleanField('barrier_open', datos.barrera_abierta)
      .booleanField('lights_on', datos.luces_parking_encendidas)
      .booleanField('alarm_active', datos.alarma_temperatura_activa);

    try {
      this.writeApi.writePoint(point);
      await this.writeApi.flush(); // fuerza el envío
      console.log(`✅ Datos almacenados en InfluxDB para dispositivo ${device_id}`);
      return true;
    } catch (error) {
      console.error('❌ Error almacenando en InfluxDB:', error);
      return false;
    }
  }
}
